File: day-08-quicksort/js/quicksort-random.js
function quicksortRandom(arr) {
  if (arr.length < 2) return arr;
  const pivotIndex = Math.floor(Math.random() * arr.length);
  const pivot = arr[pivotIndex];

  const less = [];
  const greater = [];

  for (let i = 0; i < arr.length; i++) {
    if (i === pivotIndex) continue;

    if (arr[i] <= pivot) {
      less.push(arr[i]);
    } else {
      greater.push(arr[i]);
    }
  }
  return [...quicksortRandom(less), pivot, ...quicksortRandom(greater)];
}

const myArr = [
  40, 162, 282, 410, 460, 482, 552, 565, 742, 756, 1046, 1196, 1267, 1292, 1295,
];

console.log(quicksortRandom(myArr));

File: day-08-quicksort/python/quicksort-random.py
import random

def quicksort_random(arr):
    if len(arr) < 2:
        return arr
    pivot_index = random.randint(0, len(arr) - 1)
    pivot = arr[pivot_index]
    print(pivot)

    less = [arr[i] for i in range(len(arr)) if i != pivot_index and arr[i] <= pivot]
    greater = [arr[i] for i in range(len(arr)) if i != pivot_index and arr[i] > pivot]
    return quicksort_random(less) + [pivot] + quicksort_random(greater)

print(quicksort_random([1, 2, 3, 4, 5]))
